pennylane/hf/tapering.py

```diff
@@ -0,0 +1,277 @@
+    binary_vector_indices = np.ix_(np.arange(len(pivots)), nonpivots)
+    null_vector[null_vector_indices] = -binary_matrix[binary_vector_indices] % 2
+
+    nullspace = null_vector.T
+    return nullspace
+
+
+def generate_taus(nullspace, num_qubits):
+    r"""Compute the generators :math:`\{\tau_1, \ldots, \tau_k\}` from the nullspace of
+    the binary matrix form of a Hamiltonian over the binary field :math:`\mathbb{Z}_2`.
+    These correspond to the generator set of the :math:`\mathbb{Z}_2`-symmetries present
+    in the Hamiltonian as given in `arXiv:1910.14644 <https://arxiv.org/abs/1910.14644>`_.
+
+    Args:
+        nullspace (list): kernel of the binary matrix corresponding to the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+
+    Returns:
+        generators (list): list of generators of symmetries, taus, for the Hamiltonian.
+
+    **Example**
+
+    .. code-block::
+        >>> kernel = np.array([[0, 1, 0, 0, 0, 0, 0, 0],
+                               [0, 0, 1, 1, 1, 0, 0, 0],
+                               [1, 0, 1, 0, 0, 1, 0, 0],
+                               [0, 0, 1, 0, 0, 0, 1, 0],
+                               [0, 0, 1, 1, 0, 0, 0, 1]])
+        >>> generate_taus(kernel, 4)
+         [(1.0) [X1], (1.0) [Z0 X2 X3], (1.0) [X0 Z1 X2], (1.0) [Y2], (1.0) [X2 Y3]]
+
+    """
+
+    generators = []
+    pauli_map = {"00": qml.Identity, "10": qml.PauliX, "11": qml.PauliY, "01": qml.PauliZ}
+
+    for null_vector in nullspace:
+        tau = qml.Identity(0)
+        for idx, op in enumerate(zip(null_vector[:num_qubits], null_vector[num_qubits:])):
+            x, z = op
+            tau @= pauli_map[f"{x}{z}"](idx)
+
+        ham = qml.Hamiltonian([1.0], [tau], simplify=True)
+        generators.append(ham)
+
+    return generators
+
+
+def generate_paulis(generators, num_qubits):
+    r"""Generate the single qubit Pauli X operators :math:`sigma^{x}_{i}` for each symmetries :math:`tau_j`,
+    such that it anti-commutes with :math:`tau_j` and commutes with all others symmetries :math:`tau_{k\neq j`}.
+    These are required to obtain the Clifford operators :math:`U` for the Hamiltonian :math:`H`.
+
+    Args:
+        generators (list): list of generators of symmetries, taus, for the Hamiltonian.
+        num_qubits (int): number of wires required to define the Hamiltonian.
+    Return:
+        sigma_x (list): list of single-qubit Pauli X operators which will be used to build the
+        Clifford operators `U`.
+
+    **Example**
+
+    .. code-block::
+        >>> generators = [qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(1)]),
+                          qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(2)]),
+                          qml.Hamiltonian([1.0], [qml.PauliZ(0) @ qml.PauliZ(3)])]
+        >>> generate_paulis(generators, qubits)
+         [PauliX(wires=[1]), PauliX(wires=[2]), PauliX(wires=[3])]
```

Comment:
Minor point, isn't it easier to just return the indices not the actual operators?